### PR TITLE
Prevent ref cycle creation in inner hook

### PR DESCRIPTION
**Pull request by albanD**
Towards fixing https://github.com/pytorch/pytorch/issues/82482

This PR fixes two things:

## 1) memory leak
The .detach() call prevents a true memory leak in some cases where the user function is using multiple ops in a row that save their inputs. The following chain of objects keep each other alive
- the `storage` object
- a recomputed Tensor y
- y's grad_fn FooBackward (in c++)
- FooBackward's SavedVariables (in c++)
- SavedVariable Hook
- the `inner_pack` function
- captures `storage`

Since part of this cycle is in c++, the python gc is not able to break it.
Should THPCppFunction_traverse actually visit it's SavedVariables which in turn should visit their hooks? I think the answer is yes but I haven't dived into which python object is traversing what as if there is non-unique ownership of the c++ object, it makes the traversal a lot trickier. @ezyang do you think we should dive into this more?

In this case, this can be easily solved anyways by storing `y.detach()` in the `storage` object as we don't care about the temporary backward graph that gets created during the second forward call.

## 2) Lifetime of the recomputed buffers
The new storage system is now such that the lifetime of the recomputed buffer is directly linked to the SavedVariable c++ object. Meaning that this buffer will get deleted IIF the SavedVariable is cleared.
This means that we now get the exact same behavior as the version without the saved variable hook where Tensors are saved directly on the SavedVariable object.

This is great as this solves all the cases where the non-checkpoint version used to work but the checkpoint version does not (even double access or retain_graph=True).

The one drawback of this approach though is that the buffer do NOT get cleared when the user passes in `retain_graph=True`! The next backward won't even re-run the forward as it already has all the buffers available. Is this a problem that you think we would need to find a solution for @rohan-varma or it is niche enough that we don't care for now?

### CODE REVIEW

**Comment by ezyang** [comment]
How come storage objects have reference to tensor? That seems fishy.

**Comment by ezyang** [comment]
I mean, you can try the unique reference trick to see if you can detect the cycle in python gc. But it will always be better to not create cycles in the first place. I'm not sure how to best fix this; docs are a start, but maybe we need some better way of letting users identify cycles. For example, how did you work out the cycle here? Is there something generalizable we can do?

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -350,23 +351,36 @@ def _checkpoint_without_reentrant(function, preserve_rng_state=True, *args, **kw
...
-        # TODO(varal7): Instead of returning indices, we can return things metadata (such as
+        # TODO(varal7): Instead of returning abstract object, we can return things metadata (such as
         # size, device, ...) to catch certain cases of undeterministic behavior of the forward
-        return counter - 1
+        res = Holder()
+        weak_holder_list.append(weakref.ref(res))
+        return res
+
 
     def unpack(x):
```
Comment (ezyang) [inline torch/utils/checkpoint.py:373]: so.... this is getting a Holder? Why not call the holder some sort of key object or something

**Review by ezyang — APPROVED** [review]

**Review by rohan-varma — APPROVED** [review]
Thank you for the quick fix!

**Comment by nyngwang** [comment]
@albanD: Does this fix available on the current stable version (v1.12.1)? Or will I need to install nightly for this?

**Comment by nyngwang** [comment]
> @albanD: [...] The one drawback of this approach though is that the buffer do NOT get cleared when the user passes in retain_graph=True! The next backward won't even re-run the forward as it already has all the buffers available. [...]

1. What did you mean "The next backward" here?
2. Could you elaborate more on "won't even re-run the forward"? In my use case, I divide a model into many checkpointed `nn.Module` that consists of a `nn.Sequential`. To save GPU memory each checkpointed `nn.Module` should have its trainable parameters get updated by re-run the forwardings of those layers contained in its internal `nn.Sequential`. This should reduce the GPU memory usage of the forward pass. Does this PR does anything contradict this intention?

**Comment by albanD** [comment]
> Does this fix available on the current stable version (v1.12.1)? Or will I need to install nightly for this?

It is only in nightly I'm afraid.

> What did you mean "The next backward" here?

If you do:
```
loss.backward(retain_graph=True) # This is the backward that recomputes the buffers
loss.backward() # this is the next backward
```

> Could you elaborate more on "won't even re-run the forward"?

This is only in the specific case where you run multiple backward on the same graph which is very unlikely.
If you do want that though, could you open a new issue that describes in details what you want to do?
